### PR TITLE
Preventing assignment to readonly properties of function.

### PR DESCRIPTION
**Pull request by jsdevel**
* For whatever reason in node v0.12, assigning to Function.name throws an error.
* This change now avoids assigning any name property to the return class function to avoid such an error.
* This should be entirely safe being that node would igore this anyway in v0.10.

### CODE REVIEW

File: lib/nodeJavaBridge.js [hunk 1]
```diff
@@ -223,12 +224,20 @@ java.import = function(name) {
...
       && ((modifiers & MODIFIER_STATIC) === MODIFIER_STATIC)) {
       var methodName = SyncCall(methods[i], 'getName')();
-      result[methodName + syncSuffix] = java.callStaticMethodSync.bind(java, name, methodName);
-      if (typeof asyncSuffix === 'string') {
-        result[methodName + asyncSuffix] = java.callStaticMethod.bind(java, name, methodName);
+      var syncName = methodName + syncSuffix;
+      var asyncName = methodName + asyncSuffix;
+      var promiseName = methodName + promiseSuffix;
+
+      if (readOnlyFunctionProperties.indexOf(syncName === -1)) {
```
Comment (joeferner) [inline lib/nodeJavaBridge.js]: Is the paren correct here?

File: lib/nodeJavaBridge.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ java.nativeBindingLocation = binaryPath;
 
 var syncSuffix = undefined;
 var asyncSuffix = undefined;
+var readOnlyFunctionProperties = ['name'];
```
Comment (joeferner) [inline lib/nodeJavaBridge.js]: Wouldn't setting this make all name properties read only?

Comment (joeferner) [inline lib/nodeJavaBridge.js]: Instead of this variable here, should the import function take this as an options?

**Comment by jimlloyd** [comment]
@jsdevel This bug was apparently introduced with changes I made. I'm wondering why I haven't seen this problem before. In fact it's not clear to me what triggers it. Is it specific to a platform that I don't use? I do all my development on a Mac, but continuous integration automation I use (the public Travis and an internal Jenkins server) run on Linux. Is this problem specific to Windows perhaps?

Can you provide a unit test that demonstrates the failure without the PR, and fixes the test with the PR?

**Comment by jimlloyd** [comment]
@jsdevel Sorry, I think I understand the problem now. I'm still not able to reproduce it, but I think I see how your change to the class Test is intended to trigger the problem. 

If I understand correctly, the problem will only happen with Java methods named 'name', and when a method flavor suffix is the empty string. In the defacto standard case, that would be async methods.

I'm preparing an experimental branch that will have tests for this. So far, my testing hasn't shown any errors on my Mac.

**Comment by jimlloyd** [comment]
@jsdevel: Ahh, I can reproduce this now. It only happens with *static* methods named 'name'. It doesn't happen with instance methods named 'name'.

**Comment by jimlloyd** [comment]
It seems we should use [Object.getOwnPropertyDescriptor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor) instead of checking against the fixed list readOnlyFunctionProperties.

File: lib/nodeJavaBridge.js (hunk 1 above)
Comment (jsdevel) [inline lib/nodeJavaBridge.js]: It's not!  Updating.

**Comment by jsdevel** [comment]
@jimlloyd unfortunately, `Object.getOwnPropertyDescriptor` threw errors when any of the name variables were undefined.  I feel the verbosity of the current implementation is better the the potential verbosity of `getOwnPropertyDescriptor`, but I'm open to suggestions.

@joeferner Thanks for catching the parens issue!  Fixing that allowed me to compile `Test.java` and run it successfully against `Test.class`.

File: lib/nodeJavaBridge.js (hunk 2 above)
Comment (jsdevel) [inline lib/nodeJavaBridge.js]: I'm not entirely sure what you mean.

Comment (joeferner) [inline lib/nodeJavaBridge.js]: I'm sorry, I didn't understand what this was fixing. I see now. You can ignore these two comments.

**Comment by jugglinmike** [comment]
This is an ES6 thing--Node v0.12 has an updated version of V8 which observes some of [the new semantics around Function instances' `name` property](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-function-instances-name) (which was unspecified in ES5). The property is spec'd to be configurable, so in an ideal world, we could write:

```js
delete result[methodName + syncSuffix];
result[methodName + syncSuffix] = java.callStaticMethodSync.bind(java, name, methodName);
```

...but this is not an idea world:

```
$ node -pe "Object.getOwnPropertyDescriptor(function(){}, 'name');"
{ value: '',
  writable: false,
  enumerable: false,
  configurable: false }
```

There's hope, though--recent builds of V8 get this right:

```
$ d8 -e "print(Object.getOwnPropertyDescriptor(function(){}, 'name').configurable);"
true
```

So some future release of Node.js should allow this.

In the mean time, is there any risk involved with silently ignoring properties that cannot be defined?

**Comment by jimlloyd** [comment]
@jugglinmike: 'In the mean time, is there any risk involved with silently ignoring properties that cannot be defined?'

Yes, there is risk. The bug as stands prevents any node app from importing any Java class that has a static method named `name`, even if the app never uses the method. With the proposed fix, the node app will be able to import the class, but if they try to call the method, a TypeError exception will be thrown `TypeError: string is not a function`.

One solution would be to define the function with some additional suffix, perhaps an underscore. We document this behavior, and then any application that needs to call the static method `name()`  can instead call the static method `name_()`.

(Hey @jugglinmike, I assume you're a juggler? Me too. Solid 5b, some 6b & 7b).

**Comment by jimlloyd** [comment]
@jsdevel This is looking good, but let's do the additional work I outlined in my response to @jugglinmike above. I hacked a quick experiment which seems to work. See https://github.com/RedSeal-co/node-java/commit/b037f98e7a1e4aca84c69d6d7b66d037c566c1ea

**Comment by jugglinmike** [comment]
I was almost going to suggest this, but I happen to know that @jsdevel's use case is exposing a Java API in Node with as much fidelity as possible. I'm wondering if the prefix should be configurable, since it's hard to choose one general transformation that will be acceptable for all cases.

My 2 cents regarding RedSeal-co@b037f98 is that it might be easier to read and use the code if it were factored into a "safeAttach" function. But I'm probably over-staying my welcome offering advice like that :P

(Still not solid with 5b. 4c is pretty stable, though.)

**Comment by jimlloyd** [comment]
Your 2 cents are good. As I said, my commit was a quick hack. I agree the suffix should be configurable, we just need to add another field to asyncOptions. We should decide on a default value for the suffix. I think the default can be a single underscore, but I'm happy to let it be something else.

**Comment by jsdevel** [comment]
@jimlloyd I'm not exactly sure what more you'd like me to do.  Right now, a library I'm working on is failing without this fix.

I'd like to merge this and let any cleanup / re-work happen later.

**Comment by jimlloyd** [comment]
@jsdevel Fair enough. Since @joeferner has given me commit rights I'll merge this PR now, and prepare another PR with my proposed enhancements.

**Comment by jugglinmike** [comment]
> I'd like to merge this and let any cleanup / re-work happen later.

Oh, right. I almost forgot how I found this issue in the first place. Thanks for keeping us on track, @jsdevel :)

**Comment by jimlloyd** [comment]
Yes, and thanks @jsdevel for finding and fixing the bug I introduced!

**Comment by jsdevel** [comment]
No problem @jimlloyd !  Thanks for reviewing!

**Comment by jsdevel** [comment]
@jimlloyd @joeferner can we get a version bump with this change?  I need to update my lib asap.

**Comment by jimlloyd** [comment]
I'm not an npm owner, so @joeferner will have to do the version bump and publish.

**Comment by joeferner** [comment]
published v0.5.3

**Comment by jsdevel** [comment]
Thanks @joeferner !

**Pull request merged**